source/slang/slang-ir-autodiff-transpose.h

```diff
@@ -278,6 +266,116 @@ struct DiffTransposePass
+        {
+            auto arg = fwdCall->getArg(ii);
+            
+            // If this isn't a ptr-type, make a var.
+            if (!as<IRPtrTypeBase>(arg->getDataType()) && as<IRDifferentialPairType>(arg->getDataType()))
+            {
+                auto pairType = as<IRDifferentialPairType>(arg->getDataType());
+
+                auto var = builder->emitVar(arg->getDataType());
+
+                SLANG_ASSERT(as<IRMakeDifferentialPair>(arg));
+
+                // Initialize this var to (arg.primal, 0).
+                builder->emitStore(
+                    var, 
+                    builder->emitMakeDifferentialPair(
+                        arg->getDataType(),
+                        as<IRMakeDifferentialPair>(arg)->getPrimalValue(),
+                        builder->emitCallInst(
+                            (IRType*)diffTypeContext.getDifferentialForType(builder, pairType->getValueType()),
+                            diffTypeContext.getZeroMethodForType(builder, pairType->getValueType()),
+                            List<IRInst*>())));
+                
+                args.add(var);
+                argTypes.add(builder->getInOutType(pairType));
+                isArgPtrTyped.add(false);
+            }
+            else
+            {
+                args.add(arg);
+                argTypes.add(arg->getDataType());
+                isArgPtrTyped.add(true);
+            }
+        }
+
+        args.add(revValue);
+        argTypes.add(revValue->getDataType());
+
+        auto revFnType = builder->getFuncType(argTypes, builder->getVoidType());
+        auto revCallee = builder->emitBackwardDifferentiateInst(
+            revFnType,
+            baseFn);
+
+        builder->emitCallInst(revFnType->getResultType(), revCallee, args);
+
+        List<RevGradient> gradients;
+        for (UIndex ii = 0; ii < fwdCall->getArgCount(); ii++)
+        {
+            // Is this arg relevant to auto-diff?
+            if (as<IRDifferentialPairType>(as<IRPtrTypeBase>(args[ii]->getDataType())->getValueType()))
+            {
+                // If this is ptr typed, ignore (the gradient will be accumulated on the pointer)
+                // automatically.
+                // 
+                if (!isArgPtrTyped[ii])
+                {
+                    auto diffArgType = (IRType*)diffTypeContext.getDifferentialForType(
+                        builder, 
+                        as<IRDifferentialPairType>(
+                            as<IRPtrTypeBase>(argTypes[ii])->getValueType())->getValueType());
+                    auto diffArgPtrType = builder->getPtrType(kIROp_PtrType, diffArgType);
+                    
+                    gradients.add(RevGradient(
+                        RevGradient::Flavor::Simple,
+                        fwdCall->getArg(ii),
+                        builder->emitLoad(
+                            builder->emitDifferentialPairAddressDifferential(
```

Comment:
Seeing this code I just realized that anything of pointer type ends up as an intermediary that gets reused between primal and reverse functions, we can't really store and reuse them, and will have to do recompute. Because pointers can't be materialized in HLSL/GLSL.

Reply:
The logic for generating a reverse derivative function should still be copying everything from primal, but don't use the value in the differential part of the function if there is already an intermediate available from the stored value, and then rely on DCE to remove the computations that are no longer needed due to the storage. This way we can safely preserve the computations for all intermediaries that we can't store or didn't choose to store.